File: src/HomeDarts.Domain/ThrowingRound.cs
namespace HomeDarts.Domain;

public class ThrowingRound
{
    private List<DartPoints> _dartPoints;

    public ThrowingRound() => _dartPoints = new();

    private ThrowingRound(List<DartPoints> dartPoints) => _dartPoints = dartPoints;
    
    public void Add(DartPoints points)
    {
        if (HasReachedMaxThrows())
        {
            throw new InvalidOperationException("Max. 3 throws are allowed per throwing round.");
        }
        
        _dartPoints.Add(points);
    }

    public bool HasReachedMaxThrows() => _dartPoints.Count == 3;

    public bool LastPointIsDouble()
    {
        var lastPoint = _dartPoints.LastOrDefault();

        return lastPoint?.IsDouble() ?? false;
    }

    public void Clear() => _dartPoints.Clear();
    
    public int GetTotalScore()
        => _dartPoints.Aggregate(0, (total, points) => total + points.Value);
    public ThrowingRound Copy() => new (_dartPoints.Select(x => x.Copy()).ToList());

}

File: src/React/home-darts/src/models/ThrowingRound.ts
import { DartPoints } from "./DartPoints";

export class ThrowingRound {

    private _dartPoints: DartPoints[];

    constructor() {
        this._dartPoints = [];
    }

    public get totalScore(): number {
        return this._dartPoints
            .map(dartPoints => dartPoints.value)
            .reduce((accumulator, current) => accumulator + current , 0);
    }

    public get lastPointIsDouble(): boolean {
        var lastPoint = this._dartPoints[this._dartPoints.length - 1];

        return lastPoint?.isDouble() ?? false;
    }

    public add = (points: DartPoints) : void =>
    {
        if (this.hasReachedMaxThrows())
        {
            return;
        }
        
        this._dartPoints.push(points);
    }

    public copy = () => Object.create(this);

    private hasReachedMaxThrows = () : boolean => this._dartPoints.length === 3;
}